File: spk/cli/_cmd_render.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

import os
from typing import Any
import argparse
import sys

import structlog
from colorama import Fore

import spk
import spkrs

from . import _flags

_LOGGER = structlog.get_logger("spk.cli")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    render_cmd = sub_parsers.add_parser("render", help=_render.__doc__, **parser_args)
    render_cmd.add_argument(
        "packages", metavar="PKG", nargs="+", help="The packages to resolve and render"
    )
    render_cmd.add_argument(
        "target", metavar="PATH", help="The empty directory to render into"
    )
    _flags.add_request_flags(render_cmd)
    _flags.add_solver_flags(render_cmd)
    render_cmd.set_defaults(func=_render)
    return render_cmd


def _render(args: argparse.Namespace) -> None:
    """Output the contents of an spk environment (/spfs) to a folder."""

    solver = _flags.get_solver_from_flags(args)
    for name in args.packages:
        solver.add_request(name)

    for request in _flags.parse_requests_using_flags(args, *args.packages):
        solver.add_request(request)

    try:
        generator = solver.run()
        spk.io.format_decisions(generator, sys.stdout, args.verbose)
        solution = generator.solution
    except spk.SolverError as e:
        print(spk.io.format_error(e, args.verbose), file=sys.stderr)
        sys.exit(1)

    solution = spk.build_required_packages(solution)
    stack = spk.exec.resolve_runtime_layers(solution)
    path = os.path.abspath(args.target)
    os.makedirs(path, exist_ok=True)
    if len(os.listdir(path)) != 0:
        print(
            spk.io.format_error(
                ValueError(f"Directory is not empty {path}"), args.verbose
            ),
            file=sys.stderr,
        )
        sys.exit(1)
    _LOGGER.info(f"Rendering into dir: {path}")
    spkrs.render_into_dir(stack, path)
    _LOGGER.info(f"Render completed: {path}")

File: src/cli/cmd_render.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk
use std::path::PathBuf;

use anyhow::{bail, Context, Result};
use clap::Args;

use super::{flags, Run};

/// Output the contents of an spk environment (/spfs) to a folder
#[derive(Args)]
pub struct Render {
    #[clap(flatten)]
    pub solver: flags::Solver,
    #[clap(flatten)]
    pub options: flags::Options,
    #[clap(flatten)]
    pub requests: flags::Requests,

    #[clap(short, long, global = true, parse(from_occurrences))]
    pub verbose: u32,

    #[clap(flatten)]
    pub formatter_settings: flags::DecisionFormatterSettings,

    /// The packages to resolve and render
    #[clap(name = "PKG", required = true)]
    packages: Vec<String>,

    /// The empty directory to render into
    #[clap(name = "PATH")]
    target: PathBuf,
}

#[async_trait::async_trait]
impl Run for Render {
    async fn run(&mut self) -> Result<i32> {
        let (mut solver, requests) = tokio::try_join!(
            self.solver.get_solver(&self.options),
            self.requests.parse_requests(&self.packages, &self.options)
        )?;
        for name in requests {
            solver.add_request(name);
        }

        let formatter = self.formatter_settings.get_formatter(self.verbose);
        let solution = formatter.run_and_print_resolve(&solver).await?;

        let solution = spk::build_required_packages(&solution).await?;
        let stack = spk::exec::resolve_runtime_layers(&solution).await?;
        std::fs::create_dir_all(&self.target).context("Failed to create output directory")?;
        if std::fs::read_dir(&self.target)
            .context("Failed to validate output directory")?
            .next()
            .is_some()
        {
            bail!("Output directory does not appear to be empty");
        }

        let path = self.target.canonicalize()?;
        tracing::info!("Rendering into dir: {path:?}");
        let items: Vec<String> = stack.iter().map(ToString::to_string).collect();
        let env_spec = items.join("+").parse()?;
        spfs::render_into_directory(&env_spec, &path).await?;
        tracing::info!("Render completed: {path:?}");
        Ok(0)
    }
}
